File: ios/Classes/ZendeskMessagingFlutterPlugin.swift
import Flutter
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK

public class ZendeskMessagingFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler,
    MessagingDelegate {

    private var eventSink: FlutterEventSink?

    // URL handling policy (set from Dart via setUrlPolicy). Native decides
    // synchronously; intercepted links are forwarded as a `urlClicked` event.
    private var urlPolicy: String = "sdkOpens"
    private var urlPatterns: [String] = []

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "zendesk_messaging",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "zendesk_messaging/events",
            binaryMessenger: registrar.messenger()
        )
        let instance = ZendeskMessagingFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - Method handler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        switch call.method {

        case "initialize":
            guard let channelKey = args?["channelKey"] as? String else {
                return result(FlutterError(code: "INVALID_ARGS", message: "channelKey is required", details: nil))
            }
            if Zendesk.instance != nil { return result(nil) }
            Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { [weak self] initResult in
                switch initResult {
                case .success:
                    Messaging.delegate = self
                    result(nil)
                case .failure(let error):
                    result(FlutterError(code: "INIT_FAILED", message: error.localizedDescription, details: nil))
                }
            }

        case "invalidate":
            Zendesk.invalidate()
            result(nil)

        case "getCurrentUser":
            guard let zendesk = Zendesk.instance,
                  let u = zendesk.getCurrentUser() else { return result(nil) }
            result([
                "id": u.id,
                "externalId": u.externalId,
                "authenticationType": authTypeName(u.authenticationType),
            ])

        case "loginUser":
            guard let jwt = args?["jwt"] as? String,
                  let zendesk = Zendesk.instance else {
                return result(FlutterError(code: "INVALID_ARGS", message: "jwt required / SDK not initialized", details: nil))
            }
            zendesk.loginUser(with: jwt) { loginResult in
                switch loginResult {
                case .success: result(nil)
                case .failure(let e): result(FlutterError(code: "LOGIN_FAILED", message: e.localizedDescription, details: nil))
                }
            }

        case "logoutUser":
            guard let zendesk = Zendesk.instance else { return result(nil) }
            zendesk.logoutUser { logoutResult in
                switch logoutResult {
                case .success: result(nil)
                case .failure(let e): result(FlutterError(code: "LOGOUT_FAILED", message: e.localizedDescription, details: nil))
                }
            }

        case "show":
            guard let messaging = Zendesk.instance?.messaging else {
                return result(FlutterError(code: "NOT_INITIALIZED", message: "Call initialize() first", details: nil))
            }
            let fullScreen = args?["fullScreen"] as? Bool ?? true
            let screen = buildMessagingScreen(from: args)
            let vc: UIViewController
            if let screen = screen {
                vc = messaging.messagingViewController(screen)
            } else {
                vc = messaging.messagingViewController()
            }
            DispatchQueue.main.async {
                let nav = UINavigationController(rootViewController: vc)
                nav.modalPresentationStyle = fullScreen ? .fullScreen : .pageSheet
                UIApplication.shared.topViewController?.present(nav, animated: true)
                result(nil)
            }

        case "getUnreadMessageCount":
            let count = Zendesk.instance?.messaging?.getUnreadMessageCount() ?? 0
            result(count)

        case "getUnreadMessageCountForConversation":
            guard let id = args?["conversationId"] as? String else {
                return result(FlutterError(code: "INVALID_ARGS", message: "conversationId is required", details: nil))
            }
            let count = Zendesk.instance?.messaging?.getUnreadMessageCount(conversationId: id) ?? 0
            result(count)

        case "setConversationFields":
            guard let fields = args?["fields"] as? [String: AnyHashable] else { return result(nil) }
            Zendesk.instance?.messaging?.setConversationFields(fields)
            result(nil)

        case "clearConversationFields":
            Zendesk.instance?.messaging?.clearConversationFields()
            result(nil)

        case "setConversationTags":
            guard let tags = args?["tags"] as? [String] else { return result(nil) }
            Zendesk.instance?.messaging?.setConversationTags(tags)
            result(nil)

        case "clearConversationTags":
            Zendesk.instance?.messaging?.clearConversationTags()
            result(nil)

        case "sendPageViewEvent":
            guard let pageTitle = args?["pageTitle"] as? String,
                  let url = args?["url"] as? String,
                  let zendesk = Zendesk.instance else {
                return result(FlutterError(code: "INVALID_ARGS", message: "pageTitle, url, and initialized SDK required", details: nil))
            }
            zendesk.sendPageViewEvent(PageView(pageTitle: pageTitle, url: url)) { sendResult in
                switch sendResult {
                case .success: result(nil)
                case .failure(let e): result(FlutterError(code: "PAGE_VIEW_FAILED", message: e.localizedDescription, details: nil))
                }
            }

        case "updatePushNotificationToken":
            guard let tokenHex = args?["token"] as? String else {
                return result(FlutterError(code: "INVALID_ARGS", message: "token is required", details: nil))
            }
            guard let tokenData = Data(hexString: tokenHex) else {
                return result(FlutterError(code: "INVALID_ARGS", message: "token is not valid hex", details: nil))
            }
            PushNotifications.updatePushNotificationToken(tokenData)
            result(nil)

        case "shouldBeDisplayed":
            guard let messageData = args?["messageData"] as? [AnyHashable: Any] else {
                return result("notFromMessaging")
            }
            let responsibility = PushNotifications.shouldBeDisplayed(messageData)
            result(pushResponsibilityName(responsibility))

        case "enableAnalyticsTracking":
            let enabled = args?["enabled"] as? Bool ?? true
            Zendesk.instance?.messaging?.enableInternalAnalytics(enabled: enabled)
            result(nil)

        case "setUrlPolicy":
            urlPolicy = args?["policy"] as? String ?? "sdkOpens"
            urlPatterns = args?["patterns"] as? [String] ?? []
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - MessagingDelegate

    // Called synchronously on the main thread. `true` = SDK opens the URL,
    // `false` = the app is responsible. We decide from the pre-registered
    // policy (no Dart round-trip) and, when the app is responsible, emit a
    // `urlClicked` event fire-and-forget so Dart can act on it.
    public func messaging(_ messaging: Messaging, shouldHandleURL url: URL, from source: URLSource) -> Bool {
        let appHandles: Bool
        switch urlPolicy {
        case "appHandlesAll":
            appHandles = true
        case "appHandlesMatching":
            let absolute = url.absoluteString
            appHandles = urlPatterns.contains { absolute.contains($0) }
        default:  // "sdkOpens"
            appHandles = false
        }
        if appHandles {
            let sink = eventSink
            let payload: [String: Any] = ["type": "urlClicked", "url": url.absoluteString, "source": urlSourceName(source)]
            DispatchQueue.main.async { sink?(payload) }
        }
        return !appHandles  // SDK handles the URL only if the app does not
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        guard let zendesk = Zendesk.instance else { return nil }
        zendesk.addEventObserver(self) { [weak self] event in
            guard let self else { return }
            let sink = self.eventSink
            let map = self.eventToMap(event)
            guard let map else { return }
            DispatchQueue.main.async { sink?(map) }
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        Zendesk.instance?.removeEventObserver(self)
        eventSink = nil
        return nil
    }

    // MARK: - Event mapping

    private func eventToMap(_ event: ZendeskEvent) -> [String: Any?]? {
        switch event {
        case .сonversationUnreadCountChanged(id: _, timestamp: _, data: let data):
            return [
                "type": "unreadMessageCountChanged",
                "totalUnreadCount": data.totalUnreadMessagesCount,
                "conversationId": data.conversationId,
                "unreadInConversation": data.unreadCountInConversation,
            ]
        case .authenticationFailed:
            return ["type": "authenticationFailed"]
        case .connectionStatusChanged(connectionStatus: let status):
            return ["type": "connectionStatusChanged", "status": connectionStatusName(status)]
        case .conversationAdded(conversationId: let cid):
            return ["type": "conversationAdded", "conversationId": cid]
        case .conversationStarted(id: let id, timestamp: let ts, conversationId: let cid):
            return ["type": "conversationStarted", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": cid]
        case .conversationOpened(id: let id, timestamp: let ts, conversationId: let cid):
            return ["type": "conversationOpened", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": cid as Any]
        case .messagesShown(id: let id, timestamp: let ts, conversationId: let cid, messages: let msgs):
            return [
                "type": "messagesShown",
                "id": id.uuidString,
                "timestamp": Int(ts.timeIntervalSince1970 * 1000),
                "conversationId": cid,
                "messages": msgs.map { ["id": $0.id, "role": $0.role == .user ? "user" : "business", "timestamp": Int($0.timestamp.timeIntervalSince1970 * 1000)] },
            ]
        case .sendMessageFailed(error: let e):
            return ["type": "sendMessageFailed", "message": e.localizedDescription]
        case .messagingOpened(id: let id, timestamp: let ts):
            return ["type": "messagingOpened", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000)]
        case .messagingClosed(id: let id, timestamp: let ts):
            return ["type": "messagingClosed", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000)]
        case .newConversationButtonClicked(id: let id, timestamp: let ts, data: let data):
            return ["type": "newConversationButtonClicked", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "source": newConversationSourceName(data.newConversationSource)]
        case .proactiveMessageDisplayed(id: let id, timestamp: let ts, data: let data):
            return ["type": "proactiveMessageDisplayed", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "proactiveMessageId": data.proactiveMessageId, "campaignId": data.campaignId]
        case .proactiveMessageClicked(id: let id, timestamp: let ts, data: let data):
            return ["type": "proactiveMessageClicked", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "proactiveMessageId": data.proactiveMessageId, "campaignId": data.campaignId]
        case .conversationWithAgentRequested(id: let id, timestamp: let ts, data: let data):
            return ["type": "conversationWithAgentRequested", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId]
        case .conversationAgentAssigned(id: let id, timestamp: let ts, data: let data):
            return ["type": "conversationAgentAssigned", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId]
        case .conversationServedByAgent(id: let id, timestamp: let ts, data: let data):
            return ["type": "conversationServedByAgent", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId, "agentId": data.agentId, "agentDisplayName": data.agentDisplayName, "agentMessageSource": agentMessageSourceName(data.agentMessageSource)]
        case .postbackButtonClicked(id: let id, timestamp: let ts, data: let data):
            return ["type": "postbackButtonClicked", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId, "actionName": data.actionName]
        case .conversationExtensionDisplayed(id: let id, timestamp: let ts, data: let data):
            return ["type": "conversationExtensionDisplayed", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId, "url": data.url]
        case .conversationExtensionOpened(id: let id, timestamp: let ts, data: let data):
            return ["type": "conversationExtensionOpened", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId, "url": data.url]
        case .articleClicked(id: let id, timestamp: let ts, data: let data):
            return ["type": "articleClicked", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "articleId": data.articleId, "articleTitle": data.articleTitle as Any]
        case .articleBrowserClicked(id: let id, timestamp: let ts, data: let data):
            return ["type": "articleBrowserClicked", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "url": data.url]
        case .notificationDisplayed(id: let id, timestamp: let ts, data: let data):
            return ["type": "notificationDisplayed", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId]
        case .notificationOpened(id: let id, timestamp: let ts, data: let data):
            return ["type": "notificationOpened", "id": id.uuidString, "timestamp": Int(ts.timeIntervalSince1970 * 1000), "conversationId": data.conversationId]
        case .metadataSuccess:
            return ["type": "metadataSuccess"]
        case .metadataFailure:
            return ["type": "metadataFailure"]
        default:
            return nil
        }
    }

    // MARK: - Helpers

    private func buildMessagingScreen(from args: [String: Any]?) -> MessagingScreen? {
        guard let screen = args?["screen"] as? String else { return nil }
        let exitActionStr = args?["exitAction"] as? String ?? "close"
        let exitAction: ExitAction = exitActionStr == "returnToConversationList" ? .returnToConversationList : .close
        switch screen {
        case "mostRecentConversation":
            return .showMostRecentConversation(exitAction: exitAction)
        case "conversationsList":
            return .showConversationList
        case "newConversation":
            return .showNewConversation(exitAction: exitAction)
        case "conversation":
            let cid = args?["conversationId"] as? String
            return .showConversation(conversationId: cid, exitAction: exitAction)
        default:
            return nil
        }
    }

    private func authTypeName(_ type: AuthenticationType) -> String {
        switch type {
        case .jwt: return "jwt"
        case .sessionToken: return "sessionToken"
        case .unauthenticated: return "unauthenticated"
        @unknown default: return "unauthenticated"
        }
    }

    private func connectionStatusName(_ status: ZendeskConnectionStatus) -> String {
        switch status {
        case .disconnected: return "disconnected"
        case .connected: return "connected"
        case .connectingRealtime: return "connectingRealtime"
        case .connectedRealtime: return "connectedRealtime"
        @unknown default: return "disconnected"
        }
    }

    private func pushResponsibilityName(_ r: PushResponsibility) -> String {
        switch r {
        case .messagingShouldDisplay: return "messagingShouldDisplay"
        case .messagingShouldNotDisplay: return "messagingShouldNotDisplay"
        case .notFromMessaging: return "notFromMessaging"
        @unknown default: return "notFromMessaging"
        }
    }

    private func agentMessageSourceName(_ s: AgentMessageSource) -> String {
        switch s {
        case .agentWorkspace: return "agentWorkspace"
        case .agentCopilot: return "agentCopilot"
        @unknown default: return "agentWorkspace"
        }
    }

    private func urlSourceName(_ s: URLSource) -> String {
        switch s {
        case .text: return "text"
        case .carousel: return "carousel"
        case .file: return "file"
        case .image: return "image"
        case .linkMessageAction: return "linkMessageAction"
        case .webViewMessageAction: return "webViewMessageAction"
        @unknown default: return "text"
        }
    }

    private func newConversationSourceName(_ s: NewConversationSource) -> String {
        switch s {
        case .conversationList: return "conversationList"
        @unknown default: return "conversationList"
        }
    }
}

// MARK: - UIApplication extension

private extension UIApplication {
    var topViewController: UIViewController? {
        guard let rootVC = connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap({ $0.windows })
            .first(where: { $0.isKeyWindow })?.rootViewController else { return nil }
        return findTopViewController(from: rootVC)
    }

    private func findTopViewController(from vc: UIViewController) -> UIViewController {
        if let presented = vc.presentedViewController { return findTopViewController(from: presented) }
        if let nav = vc as? UINavigationController, let top = nav.topViewController { return findTopViewController(from: top) }
        if let tab = vc as? UITabBarController, let selected = tab.selectedViewController { return findTopViewController(from: selected) }
        return vc
    }
}

// MARK: - Data hex extension

private extension Data {
    /// Parses an even-length, fully hexadecimal string. Returns `nil` for any
    /// malformed input rather than silently forwarding a truncated token.
    init?(hexString: String) {
        let clean = hexString.replacingOccurrences(of: " ", with: "")
        guard !clean.isEmpty, clean.count % 2 == 0 else { return nil }
        var data = Data(capacity: clean.count / 2)
        var index = clean.startIndex
        while index < clean.endIndex {
            let next = clean.index(index, offsetBy: 2)
            guard let byte = UInt8(clean[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        self = data
    }
}

File: android/src/main/kotlin/com/zendesk/zendesk_messaging_flutter/ZendeskMessagingFlutterPlugin.kt
package com.zendesk.zendesk_messaging_flutter

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import zendesk.android.Zendesk
import zendesk.android.events.ZendeskEvent
import zendesk.android.events.ZendeskEventListener
import zendesk.android.messaging.MessagingDelegate
import zendesk.android.messaging.UrlSource
import zendesk.android.pageviewevents.PageView
import zendesk.messaging.android.DefaultMessagingFactory
import zendesk.messaging.android.push.PushNotifications
import zendesk.android.messaging.MessagingScreen as ZDKMessagingScreen

class ZendeskMessagingFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var applicationContext: Context? = null
    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    // URL handling policy (set from Dart via setUrlPolicy). Native decides
    // synchronously; intercepted links are forwarded as a `urlClicked` event.
    private var urlPolicy: String = "sdkOpens"
    private var urlPatterns: List<String> = emptyList()

    companion object {
        private var initialized = false
    }

    private val eventListener = ZendeskEventListener { event ->
        val map: Map<String, Any?> = when (event) {
            is ZendeskEvent.UnreadMessageCountChanged -> mapOf(
                "type" to "unreadMessageCountChanged",
                "totalUnreadCount" to event.data.totalUnreadMessagesCount,
                "conversationId" to event.data.conversationId,
                "unreadInConversation" to event.data.unreadCountInConversation,
            )
            is ZendeskEvent.AuthenticationFailed -> mapOf("type" to "authenticationFailed")
            is ZendeskEvent.FieldValidationFailed -> mapOf("type" to "fieldValidationFailed")
            is ZendeskEvent.ConnectionStatusChanged -> mapOf(
                "type" to "connectionStatusChanged",
                "status" to enumToCamel(event.connectionStatus.name),
            )
            is ZendeskEvent.ConversationAdded -> mapOf(
                "type" to "conversationAdded",
                "conversationId" to event.conversationId,
            )
            is ZendeskEvent.ConversationStarted -> mapOf(
                "type" to "conversationStarted",
                "id" to event.id.toString(),
                "conversationId" to event.conversationId,
                "timestamp" to event.timestamp,
            )
            is ZendeskEvent.ConversationOpened -> mapOf(
                "type" to "conversationOpened",
                "id" to event.id.toString(),
                "conversationId" to event.conversationId,
                "timestamp" to event.timestamp,
            )
            is ZendeskEvent.MessagesShown -> mapOf(
                "type" to "messagesShown",
                "id" to event.id.toString(),
                "conversationId" to event.conversationId,
                "timestamp" to event.timestamp,
                "messages" to event.messages.map {
                    mapOf("id" to it.id, "role" to enumToCamel(it.role.name), "timestamp" to it.timestamp)
                },
            )
            is ZendeskEvent.SendMessageFailed -> mapOf(
                "type" to "sendMessageFailed",
                "message" to event.cause.message,
            )
            is ZendeskEvent.MessagingOpened -> mapOf(
                "type" to "messagingOpened",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
            )
            is ZendeskEvent.MessagingClosed -> mapOf(
                "type" to "messagingClosed",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
            )
            is ZendeskEvent.NewConversationButtonClicked -> mapOf(
                "type" to "newConversationButtonClicked",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "source" to enumToCamel(event.data.newConversationSource.name),
            )
            is ZendeskEvent.ProactiveMessageDisplayed -> mapOf(
                "type" to "proactiveMessageDisplayed",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "proactiveMessageId" to event.data.proactiveMessageId.toString(),
                "campaignId" to event.data.campaignId,
            )
            is ZendeskEvent.ProactiveMessageClicked -> mapOf(
                "type" to "proactiveMessageClicked",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "proactiveMessageId" to event.data.proactiveMessageId.toString(),
                "campaignId" to event.data.campaignId,
            )
            is ZendeskEvent.ConversationWithAgentRequested -> mapOf(
                "type" to "conversationWithAgentRequested",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
            )
            is ZendeskEvent.ConversationAgentAssigned -> mapOf(
                "type" to "conversationAgentAssigned",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
            )
            is ZendeskEvent.ConversationServedByAgent -> mapOf(
                "type" to "conversationServedByAgent",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
                "agentId" to event.data.agentId,
                "agentDisplayName" to event.data.agentDisplayName,
                "agentMessageSource" to enumToCamel(event.data.agentMessageSource.name),
            )
            is ZendeskEvent.PostbackButtonClicked -> mapOf(
                "type" to "postbackButtonClicked",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
                "actionName" to event.data.actionName,
            )
            is ZendeskEvent.ConversationExtensionOpened -> mapOf(
                "type" to "conversationExtensionOpened",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
                "url" to event.data.url,
            )
            is ZendeskEvent.ConversationExtensionDisplayed -> mapOf(
                "type" to "conversationExtensionDisplayed",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
                "url" to event.data.url,
            )
            is ZendeskEvent.ArticleClicked -> mapOf(
                "type" to "articleClicked",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "articleId" to event.data.articleId.toString(),
                "articleTitle" to event.data.articleTitle,
            )
            is ZendeskEvent.ArticleBrowserClicked -> mapOf(
                "type" to "articleBrowserClicked",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "url" to event.data.url,
            )
            is ZendeskEvent.NotificationDisplayed -> mapOf(
                "type" to "notificationDisplayed",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
            )
            is ZendeskEvent.NotificationOpened -> mapOf(
                "type" to "notificationOpened",
                "id" to event.id.toString(),
                "timestamp" to event.timestamp,
                "conversationId" to event.data.conversationId,
            )
            is ZendeskEvent.MetadataSuccess -> mapOf("type" to "metadataSuccess")
            is ZendeskEvent.MetadataFailure -> mapOf("type" to "metadataFailure")
            else -> return@ZendeskEventListener
        }
        mainHandler.post { eventSink?.success(map) }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "zendesk_messaging")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "zendesk_messaging/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        // Drop the process-global delegate so a torn-down plugin isn't retained/invoked.
        zendesk.android.messaging.Messaging.setDelegate(null)
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() { activity = null }
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) { activity = binding.activity }
    override fun onDetachedFromActivityForConfigChanges() { activity = null }

    @Suppress("UNCHECKED_CAST")
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            "initialize" -> {
                val channelKey = call.argument<String>("channelKey")
                    ?: return result.error("INVALID_ARGS", "channelKey is required", null)
                val ctx = applicationContext
                    ?: return result.error("NO_CONTEXT", "Plugin not attached to engine", null)
                if (initialized) return result.success(null)

                Zendesk.initialize(
                    context = ctx,
                    channelKey = channelKey,
                    successCallback = {
                        initialized = true
                        installMessagingDelegate()
                        result.success(null)
                    },
                    failureCallback = { result.error("INIT_FAILED", it.message, null) },
                    messagingFactory = DefaultMessagingFactory()
                )
            }

            "invalidate" -> {
                Zendesk.invalidate()
                initialized = false
                result.success(null)
            }

            "getCurrentUser" -> {
                Zendesk.instance.getCurrentUser(
                    successCallback = { user ->
                        result.success(user?.let {
                            mapOf(
                                "id" to it.id,
                                "externalId" to it.externalId,
                                "authenticationType" to enumToCamel(it.authenticationType.name),
                            )
                        })
                    },
                    // no failure callback variant — just return null on error
                )
            }

            "loginUser" -> {
                val jwt = call.argument<String>("jwt")
                    ?: return result.error("INVALID_ARGS", "jwt is required", null)
                Zendesk.instance.loginUser(
                    jwt = jwt,
                    successCallback = { result.success(null) },
                    failureCallback = { result.error("LOGIN_FAILED", it.message, null) }
                )
            }

            "logoutUser" -> {
                Zendesk.instance.logoutUser(
                    successCallback = { result.success(null) },
                    failureCallback = { result.error("LOGOUT_FAILED", it.message, null) }
                )
            }

            "show" -> {
                val act = activity ?: return result.error("NO_ACTIVITY", "No activity available", null)
                val screen = buildMessagingScreen(call)
                if (screen != null) {
                    Zendesk.instance.messaging.showMessaging(act, screen)
                } else {
                    Zendesk.instance.messaging.showMessaging(act)
                }
                result.success(null)
            }

            "getUnreadMessageCount" -> {
                result.success(Zendesk.instance.messaging.getUnreadMessageCount())
            }

            "getUnreadMessageCountForConversation" -> {
                val id = call.argument<String>("conversationId")
                    ?: return result.error("INVALID_ARGS", "conversationId is required", null)
                result.success(Zendesk.instance.messaging.getUnreadMessageCount(id))
            }

            "setConversationFields" -> {
                val fields = call.argument<Map<String, Any>>("fields") ?: emptyMap()
                Zendesk.instance.messaging.setConversationFields(fields)
                result.success(null)
            }

            "clearConversationFields" -> {
                Zendesk.instance.messaging.clearConversationFields()
                result.success(null)
            }

            "setConversationTags" -> {
                val tags = call.argument<List<String>>("tags") ?: emptyList()
                Zendesk.instance.messaging.setConversationTags(tags)
                result.success(null)
            }

            "clearConversationTags" -> {
                Zendesk.instance.messaging.clearConversationTags()
                result.success(null)
            }

            "sendPageViewEvent" -> {
                val pageTitle = call.argument<String>("pageTitle")
                    ?: return result.error("INVALID_ARGS", "pageTitle is required", null)
                val url = call.argument<String>("url")
                    ?: return result.error("INVALID_ARGS", "url is required", null)
                Zendesk.instance.sendPageView(
                    pageView = PageView(url = url, pageTitle = pageTitle),
                    successCallback = { result.success(null) },
                    failureCallback = { result.error("PAGE_VIEW_FAILED", it.message, null) }
                )
            }

            "updatePushNotificationToken" -> {
                val token = call.argument<String>("token")
                    ?: return result.error("INVALID_ARGS", "token is required", null)
                PushNotifications.updatePushNotificationToken(token)
                result.success(null)
            }

            "shouldBeDisplayed" -> {
                val messageData = call.argument<Map<String, String>>("messageData") ?: emptyMap()
                val responsibility = PushNotifications.shouldBeDisplayed(messageData)
                result.success(responsibility.name)
            }

            "displayNotification" -> {
                val ctx = applicationContext
                    ?: return result.error("NO_CONTEXT", "Plugin not attached to engine", null)
                val messageData = call.argument<Map<String, String>>("messageData") ?: emptyMap()
                PushNotifications.displayNotification(ctx, messageData)
                result.success(null)
            }

            "setNotificationSmallIconResourceName" -> {
                val resourceName = call.argument<String>("resourceName")
                    ?: return result.error("INVALID_ARGS", "resourceName is required", null)
                val ctx = applicationContext
                    ?: return result.error("NO_CONTEXT", "Plugin not attached to engine", null)
                val resId = ctx.resources.getIdentifier(resourceName, "drawable", ctx.packageName)
                PushNotifications.setNotificationSmallIconId(if (resId != 0) resId else null)
                result.success(null)
            }

            "enableAnalyticsTracking" -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                Zendesk.instance.messaging.enableAnalyticsTracking(enabled)
                result.success(null)
            }

            "setUrlPolicy" -> {
                urlPolicy = call.argument<String>("policy") ?: "sdkOpens"
                urlPatterns = call.argument<List<String>>("patterns") ?: emptyList()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun buildMessagingScreen(call: MethodCall): ZDKMessagingScreen? {
        val screen = call.argument<String>("screen") ?: return null
        val exitActionStr = call.argument<String>("exitAction") ?: "close"
        val exitAction = if (exitActionStr == "returnToConversationList")
            ZDKMessagingScreen.ExitAction.ReturnToConversationList
        else
            ZDKMessagingScreen.ExitAction.Close
        return when (screen) {
            "mostRecentConversation" -> ZDKMessagingScreen.MostRecentActiveConversation(onExit = exitAction)
            "conversationsList" -> ZDKMessagingScreen.ConversationsList
            "newConversation" -> ZDKMessagingScreen.NewConversation(onExit = exitAction)
            "conversation" -> {
                val id = call.argument<String>("conversationId") ?: return null
                ZDKMessagingScreen.Conversation(id = id, onExit = exitAction)
            }
            else -> null
        }
    }

    private fun installMessagingDelegate() {
        zendesk.android.messaging.Messaging.setDelegate(object : MessagingDelegate() {
            // Called synchronously on the main thread. Return true = SDK opens the
            // URL, false = the app is responsible. We decide from the pre-registered
            // policy (no Dart round-trip) and, when the app is responsible, emit a
            // `urlClicked` event fire-and-forget so Dart can act on it.
            override fun shouldHandleUrl(url: String, urlSource: UrlSource): Boolean {
                val appHandles = when (urlPolicy) {
                    "appHandlesAll" -> true
                    "appHandlesMatching" -> urlPatterns.any { url.contains(it) }
                    else -> false  // "sdkOpens"
                }
                if (appHandles) {
                    val payload = mapOf(
                        "type" to "urlClicked",
                        "url" to url,
                        "source" to urlSourceName(urlSource),
                    )
                    mainHandler.post { eventSink?.success(payload) }
                }
                return !appHandles  // SDK handles the URL only if the app does not
            }
        })
    }

    private fun urlSourceName(source: UrlSource): String = when (source) {
        UrlSource.TEXT -> "text"
        UrlSource.CAROUSEL -> "carousel"
        UrlSource.FILE -> "file"
        UrlSource.IMAGE -> "image"
        UrlSource.LINK_MESSAGE_ACTION -> "linkMessageAction"
        UrlSource.WEBVIEW_MESSAGE_ACTION -> "webViewMessageAction"
    }

    // Normalizes an SDK enum constant (SCREAMING_SNAKE_CASE) to the exact
    // camelCase string the iOS side emits and the Dart layer expects — e.g.
    // CONNECTING_REALTIME -> "connectingRealtime", SESSION_TOKEN -> "sessionToken".
    // (UrlSource keeps its own explicit mapping above for the WebView casing.)
    private fun enumToCamel(name: String): String {
        val parts = name.lowercase().split("_")
        return parts.first() + parts.drop(1).joinToString("") { part ->
            part.replaceFirstChar { it.uppercaseChar() }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        Zendesk.instance.addEventListener(eventListener)
    }

    override fun onCancel(arguments: Any?) {
        Zendesk.instance.removeEventListener(eventListener)
        eventSink = null
    }
}
